### DISCUSSION TITLE
How can I install rust compiler using CIBW_BEFORE_ALL ?

### DISCUSSION THREAD

**Discussion by shi-yan** [pypa/cibuildwheel#1018] [Q&A]
My python module is written in rust and I want to use cibuildwheel to build it

But since the manylinux docker image doesn't have rust compiler, I'm trying to install it before building:

```
CIBW_BEFORE_ALL=curl https://sh.rustup.rs -sSf | sh -s -- -y && ls $HOME/.cargo/ && export PATH="$HOME/.cargo/bin:$PATH"
```

this doesn't work, even installation was fine, I couldn't set the environment variable $PATH, because I think cibuildwheel run the above using another sh instance:

```
sh -c 'curl https://sh.rustup.rs -sSf | sh -s -- -y && ls $HOME/.cargo/ && export PATH="$HOME/.cargo/bin:$PATH"'
```

once the above process is finished, all set environment variables will be gone.

this results in the following error:

```
Building wheel...

    + rm -rf /tmp/cibuildwheel/built_wheel
    + mkdir -p /tmp/cibuildwheel/built_wheel
    + python -m pip wheel /project --wheel-dir=/tmp/cibuildwheel/built_wheel --no-deps
  ERROR: Command errored out with exit status 1:
   command: /opt/python/cp36-cp36m/bin/python /opt/python/cp36-cp36m/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmpyo5_pt3m
       cwd: /project
  Complete output (6 lines):
  
  Cargo, the Rust package manager, is not installed or is not on PATH.
  This package requires Rust and Cargo to compile extensions. Install it through
  the system's package manager or via https://rustup.rs/
  
  Checking for Rust toolchain....
  ----------------------------------------
```

how should I resolve this?

**Comment by joerick** [comment]
You can't export an environment variable from `before_all` because it runs in a subprocess to cibuildwheel.

To add an environment variable for the build, you should use `CIBW_ENVIRONMENT` instead, like: 

```
CIBW_ENVIRONMENT='PATH="$HOME/.cargo/bin:$PATH"'
```
